components/webcompat/webcompat_exceptions_service.cc

```diff
@@ -0,0 +1,192 @@
+  } else {
+    return false;
+  }
+}
+
+void WebcompatExceptionsService::AddRules(
+    const base::Value::List& include_strings,
+    const base::Value::Dict& rule_dict) {
+  const base::Value* exceptions = rule_dict.Find(kExceptions);
+  if (exceptions->is_list()) {
+    for (const base::Value& include_string : include_strings) {
+      const auto pattern =
+          ContentSettingsPattern::FromString(include_string.GetString());
+      for (const base::Value& exception : exceptions->GetList()) {
+        const bool success = AddRule(pattern, exception.GetString());
+        if (!success) {
+          DLOG(ERROR) << "Unrecognized webcompat exception "
+                      << exception.GetString();
+        }
+      }
+    }
+  } else {
+    DLOG(ERROR) << "Malformed exceptions list in "
+                << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+  }
+}
+
+const std::vector<ContentSettingsPattern>&
+WebcompatExceptionsService::GetPatterns(ContentSettingsType webcompat_type) {
+  if (patterns_by_webcompat_type_.contains(webcompat_type)) {
+    return patterns_by_webcompat_type_.at(webcompat_type);
+  } else {
+    return kEmptyPatternVector;
+  }
+}
+
+void WebcompatExceptionsService::OnJsonFileDataReady(
+    const std::string& contents) {
+  if (contents.empty()) {
+    // We don't have the file yet.
+    return;
+  }
+  const auto json_root = base::JSONReader::Read(contents);
+  if (json_root == std::nullopt) {
+    DLOG(ERROR) << "Failed to parse " << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+    return;
+  }
+  if (!json_root->is_list()) {
+    DLOG(ERROR) << "Didn't find expected list in "
+                << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+    return;
+  }
+  for (const auto& rule : json_root->GetList()) {
+    if (!rule.is_dict()) {
+      // Something is wrong with the rule definition; skip it.
+      DLOG(ERROR) << "Found a malformed rule in "
+                  << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+      continue;
+    }
+    const auto& rule_dict = rule.GetDict();
+    const auto* include = rule_dict.Find(kInclude);
+    if (include->is_list()) {
+      AddRules(include->GetList(), rule_dict);
+    } else if (include->is_string()) {
+      DLOG(ERROR) << "Not implemented yet";
+    } else {
+      DLOG(ERROR) << "Malformed include attribute in "
+                  << WEBCOMPAT_EXCEPTIONS_JSON_FILE;
+    }
+  }
+}
+
+// implementation of LocalDataFilesObserver
+void WebcompatExceptionsService::OnComponentReady(
+    const std::string& component_id,
+    const base::FilePath& install_dir,
+    const std::string& manifest) {
+  LoadWebcompatExceptions(install_dir);
+}
+
+WebcompatExceptionsService::~WebcompatExceptionsService() {}
+
+// static
+WebcompatExceptionsService* WebcompatExceptionsService::CreateInstance(
+    LocalDataFilesService* local_data_files_service) {
+  singleton = new WebcompatExceptionsService(local_data_files_service);
+  return singleton;
+}
+
+// static
+WebcompatExceptionsService* WebcompatExceptionsService::GetInstance() {
```

Comment:
i wonder if we can combine Get and Create into a single method? It feels like this is a more common pattern

Reply:
In this case, I'm not sure it makes sense because we need the `local_data_files_service` to create the instance but we don't have it when we need to get the instance (in the `RemoteListProvider`).